File: tests/functional/Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/messageWithAttachment/messageWithAttachmentDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory, InputHints } = require('botbuilder');
const { ComponentDialog, ChoicePrompt, WaterfallDialog, ChoiceFactory, DialogTurnStatus, ConfirmPrompt } = require('botbuilder-dialogs');
const fs = require('fs');
const path = require('path');

const ATTACHMENT_TYPE_PROMPT = 'AttachmentTypePrompt';
const CONFIRM_PROMPT = 'ConfirmPrompt';
const WATERFALL_DIALOG = 'WaterfallDialog';

class MessageWithAttachmentDialog extends ComponentDialog {
  /**
   * @param {string} dialogId
   * @param {string} serverUrl
   */
  constructor (dialogId, serverUrl) {
    super(dialogId);

    this.picture = 'architecture-resize.png';
    this.serverUrl = serverUrl;

    this.addDialog(new ChoicePrompt(ATTACHMENT_TYPE_PROMPT))
      .addDialog(new ConfirmPrompt(CONFIRM_PROMPT))
      .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
        this.selectAttachmentType.bind(this),
        this.sendActivityWithAttachment.bind(this),
        this.finalStep.bind(this)
      ]));

    this.initialDialogId = WATERFALL_DIALOG;
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async selectAttachmentType (stepContext) {
    const messageText = 'What attachment type do you want?';
    const repromptMessageText = 'That was not a valid choice, please select a valid card type.';
    const options = {
      prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
      retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
      choices: ChoiceFactory.toChoices(['Inline', 'Internet'])
    };

    return stepContext.prompt(ATTACHMENT_TYPE_PROMPT, options);
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async sendActivityWithAttachment (stepContext) {
    const attachmentType = stepContext.result.value.toLowerCase();
    const reply = MessageFactory.text('', '', InputHints.IgnoringInput);

    switch (attachmentType) {
      case 'inline':
        reply.text = 'This is an inline attachment.';
        reply.attachments = [this.getInlineAttachment()];
        break;

      case 'internet':
        reply.text = 'This is an attachment from a HTTP URL.';
        reply.attachments = [this.getInternetAttachment()];
        break;

      default:
        throw new Error(`Invalid card type ${attachmentType}`);
    }

    await stepContext.context.sendActivity(reply);

    // Ask to submit another or end.
    const messageText = 'Do you want another type of attachment?';
    const repromptMessageText = "That's an invalid choice.";

    return stepContext.prompt(CONFIRM_PROMPT, {
      prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
      retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput)
    });
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async finalStep (stepContext) {
    const tryAnother = stepContext.result;

    if (tryAnother) {
      return stepContext.replaceDialog(this.initialDialogId);
    }

    return { status: DialogTurnStatus.complete };
  }

  /**
   * Returns an inline attachment.
   * @returns {import('botbuilder').Attachment}
   */
  getInlineAttachment () {
    const filepath = path.resolve(process.cwd(), 'images', this.picture);
    const file = fs.readFileSync(filepath, 'base64');
    return {
      name: `Files/${this.picture}`,
      contentType: 'image/png',
      contentUrl: `data:image/png;base64,${file}`
    };
  }

  /**
   * Returns an attachment to be sent to the user from a HTTPS URL.
   * @returns {import('botbuilder').Attachment}
   */
  getInternetAttachment () {
    return {
      name: `Files/${this.picture}`,
      contentType: 'image/png',
      contentUrl: `${this.serverUrl}/images/architecture-resize.png`
    };
  }
}

module.exports.MessageWithAttachmentDialog = MessageWithAttachmentDialog;

File: tests/functional/Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/MessageWithAttachment/MessageWithAttachmentDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Schema;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.MessageWithAttachment
{
    public class MessageWithAttachmentDialog : ComponentDialog
    {
        private const string Picture = "architecture-resize.png";
        private readonly Uri _serverUrl;

        public MessageWithAttachmentDialog(Uri serverUrl)
            : base(nameof(MessageWithAttachmentDialog))
        {
            _serverUrl = serverUrl;
            AddDialog(new ChoicePrompt(nameof(ChoicePrompt)));
            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { SelectAttachmentTypeAsync, SendActivityWithAttachmentAsync, FinalStepAsync }));
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> SelectAttachmentTypeAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            const string messageText = "What attachment type do you want?";
            const string repromptMessageText = "That was not a valid choice, please select a valid card type.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Choices = new List<Choice>
                {
                    new Choice("Inline"),
                    new Choice("Internet")
                }
            };

            // Ask the user to enter their name.
            return await stepContext.PromptAsync(nameof(ChoicePrompt), options, cancellationToken);
        }

        private async Task<DialogTurnResult> SendActivityWithAttachmentAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var attachmentType = ((FoundChoice)stepContext.Result).Value.ToLowerInvariant();
            var reply = new Activity(ActivityTypes.Message) { InputHint = InputHints.IgnoringInput };
            switch (attachmentType)
            {
                case "inline":
                    reply.Text = "This is an inline attachment.";
                    reply.Attachments = new List<Attachment> { GetInlineAttachment() };
                    break;

                case "internet":
                    reply.Text = "This is an attachment from a HTTP URL.";
                    reply.Attachments = new List<Attachment> { GetInternetAttachment() };
                    break;

                default:
                    throw new InvalidOperationException($"Invalid card type {attachmentType}");
            }

            await stepContext.Context.SendActivityAsync(reply, cancellationToken);

            // Ask to submit another or end.
            const string messageText = "Do you want another type of attachment?";
            const string repromptMessageText = "That's an invalid choice.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            };

            return await stepContext.PromptAsync(nameof(ConfirmPrompt), options, cancellationToken);
        }

        private async Task<DialogTurnResult> FinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tryAnother = (bool)stepContext.Result;
            if (tryAnother)
            {
                return await stepContext.ReplaceDialogAsync(InitialDialogId, cancellationToken: cancellationToken);
            }

            return new DialogTurnResult(DialogTurnStatus.Complete);
        }

        private Attachment GetInlineAttachment()
        {
            var imagePath = Path.Combine(Environment.CurrentDirectory, "wwwroot", "images", Picture);
            var imageData = Convert.ToBase64String(File.ReadAllBytes(imagePath));

            return new Attachment
            {
                Name = $"Files/{Picture}",
                ContentType = "image/png",
                ContentUrl = $"data:image/png;base64,{imageData}",
            };
        }

        private Attachment GetInternetAttachment()
        {
            return new Attachment
            {
                Name = $"Files/{Picture}",
                ContentType = "image/png",
                ContentUrl = $"{_serverUrl}images/{Picture}",
            };
        }
    }
}
